### PR TITLE
refactor: move protobuf SourceTree to its own BUILD target

### PR DESCRIPTION
**Pull request by justbuchanan**
This class is internally replaced by a google3-specific implementation. Isolating it into its own
BUILD target makes it easier to swap out.

### CODE REVIEW

File: kythe/cxx/indexer/proto/kythe_indexer_main.cc [hunk 1]
```diff
@@ -28,6 +28,7 @@
 #include <string>
 
 #include "absl/strings/match.h"
+#include "absl/strings/str_cat.h"
```
Comment (salguarnieri) [inline kythe/cxx/indexer/proto/kythe_indexer_main.cc]: Was this missing before?

File: kythe/cxx/indexer/proto/indexer_frontend.cc [hunk 2]
```diff
@@ -20,128 +20,19 @@
...
 #include "kythe/cxx/common/file_vname_generator.h"
 #include "kythe/cxx/common/indexing/KytheGraphRecorder.h"
 #include "kythe/cxx/common/path_utils.h"
 #include "kythe/cxx/indexer/proto/proto_analyzer.h"
+#include "kythe/cxx/indexer/proto/source_tree.h"
 #include "kythe/proto/analysis.pb.h"
 
+namespace kythe {
+
 using absl::StrSplit;
```
Comment (salguarnieri) [inline kythe/cxx/indexer/proto/indexer_frontend.cc]: Why place these using statements inside the namespace here (but not in the new sourcetree file)?

**Comment by salguarnieri** [comment]
I added @shahms for C++

File: kythe/cxx/indexer/proto/indexer_frontend.cc (hunk 2 above)
Comment (justbuchanan) [reply kythe/cxx/indexer/proto/indexer_frontend.cc]: Thanks - I moved the ones in sourcetree.cc inside the namespace too.

File: kythe/cxx/indexer/proto/kythe_indexer_main.cc (hunk 1 above)
Comment (justbuchanan) [reply kythe/cxx/indexer/proto/kythe_indexer_main.cc]: It was - we were getting that #include for free through something that was transitively included by indexer_frontend.h.

File: kythe/cxx/indexer/proto/source_tree.cc [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+using std::pair;
+using std::vector;
+
+namespace {
+
+// TODO(justbuchanan): why isn't there a replace_all=false version of
+// StrReplace() in open-source abseil?
+/// Finds the first occurrence of @oldsub in @s and replaces it with @newsub. If
+/// @oldsub is not present, just returns @s.
+std::string StringReplaceFirst(absl::string_view s, absl::string_view oldsub,
```
[the remark this answers is not in the document]
Comment (justbuchanan) [reply kythe/cxx/indexer/proto/source_tree.cc]: Much better, thanks!

File: kythe/cxx/indexer/proto/source_tree.cc [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+}  // namespace
+
+bool PreloadedProtoFileTree::AddFile(const std::string& filename,
+                                     const std::string& contents) {
+  VLOG(1) << filename << " added to PreloadedProtoFileTree";
+  return InsertIfNotPresent(&file_map_, filename, contents);
+}
+
+google::protobuf::io::ZeroCopyInputStream* PreloadedProtoFileTree::Open(
```
[the remark this answers is not in the document]
Comment (justbuchanan) [reply kythe/cxx/indexer/proto/source_tree.cc]: Agreed, but sadly this method is part of proto's `SourceTree` interface that we're implementing: https://github.com/protocolbuffers/protobuf/blob/b6311767e61c77f0adaea3c389d7a2963201c75d/src/google/protobuf/compiler/importer.h#L224.

[the remark this answers is not in the document]
Comment (shahms) [reply kythe/cxx/indexer/proto/source_tree.cc]: Ah, yeah, fair enough :-)

**Comment by justbuchanan** [comment]
Thanks for the reviews!

**Pull request merged**